### PR TITLE
Fix alloc in libsodium/test_crypto_generichash_multipart using MaybeUninit

### PR DESCRIPTION
**Pull request by kpp**
Fixes https://github.com/sodiumoxide/sodiumoxide/issues/338

### CODE REVIEW

**Comment by kpp** [comment]
@RalfJung please review. Is it safe?

**Comment by Dylan-DPC** [comment]
let's do it the official way. Pinging @rust-secure-code for a review (hope this works)

**Comment by RalfJung** [comment]
This looks reasonable, but I have not checked whether `crypto_generichash_init` and friends are actually okay when being called with a pointer to uninitialized memory.

**Comment by kpp** [comment]
crypto_generichash_init is okay when being called with a pointer to uninitialized memory. crypto_generichash_update and crypto_generichash_final are not

File: libsodium-sys/tests/crypto/crypto_generichash.rs [hunk 1]
```diff
@@ -95,26 +95,25 @@ fn test_crypto_generichash_multipart() {
...
 
     assert_eq!(
-        unsafe { crypto_generichash_init(pst, key.as_ptr(), key.len(), out.len()) },
+        unsafe { crypto_generichash_init(pst.as_mut_ptr(), key.as_ptr(), key.len(), out.len()) },
         0
     );
 
     assert_eq!(
-        unsafe { crypto_generichash_update(pst, m.as_ptr(), m.len() as u64) },
+        unsafe { crypto_generichash_update(pst.as_mut_ptr(), m.as_ptr(), m.len() as u64) },
```
Comment (dnaq) [inline libsodium-sys/tests/crypto/crypto_generichash.rs]: Needs an ‘assume_init()’ before this line

Comment (kpp) [reply libsodium-sys/tests/crypto/crypto_generichash.rs]: Why?

**Comment by RalfJung** [comment]
> crypto_generichash_init is okay when being called with a pointer to uninitialized memory. crypto_generichash_update and crypto_generichash_final are not

I suppose `crypto_generichash_init` initializes the memory though, and then things are okay?

File: libsodium-sys/tests/crypto/crypto_generichash.rs (hunk 1 above)
Comment (RalfJung) [reply libsodium-sys/tests/crypto/crypto_generichash.rs]: I agree, no reason to `assume_init` here.

**Comment by kpp** [comment]
> and then things are okay?

Yes

File: libsodium-sys/tests/crypto/crypto_generichash.rs (hunk 1 above)
Comment (dnaq) [reply libsodium-sys/tests/crypto/crypto_generichash.rs]: Memory is initialized after the call to ‘generichash_init()’ and we should strive to always be clear in our code.

Comment (kpp) [reply libsodium-sys/tests/crypto/crypto_generichash.rs]: Fixed.

**Comment by kpp** [comment]
bors: r+

**Pull request merged**